### ISSUE TITLE
Ensure that ASN.1 conversion is entirely deterministic

### ISSUE THREAD

**Issue by awwad** [uptane/obsolete-reference-implementation#58]
It is important that the conversion from the internal formats used by Uptane and TUF to ASN.1/DER be predictable because they are signed over their hash and signatures must be reliably validated.

In particular, we should spot check to make sure that things like generating lists from dictionaries always produce sorted order. (Fixing that now for ECU Manifests)

**Comment by JustinCappos** [comment]
Doesn't DER ensure predictable encodings?  By definition, DER provides
exactly one way to encode an ASN.1 value.

**Comment by awwad** [comment]
Yes - sorry, it occurs to me that the title and description were misleading. Clarified both.

In short: conversion to DER from the same ASN.1 is deterministic. The issue here is with conversion into ASN.1 from our internal format, which was not, in one particular way.

Here, I'm calling for a broader audit of these conversions into pyasn1 with this particular problem in mind.

**Comment by awwad** [comment]
I had to make similar changes to a variety of the ASN.1 converters in Uptane and the TUF fork to resolve more of these issues, because it kept breaking things in the demo when multiple target files or hashes were present.

- [Commit handling TUF targets](https://github.com/awwad/tuf/commit/e612b6b8399125c3b416b68091812110b4fc1ea6)
- [Commit handling TUF snapshot](https://github.com/awwad/tuf/commit/a443e5a66115cb5270d01f8352147582c450060a)
- [Commit handling Uptane Vehicle Manifests](https://github.com/uptane/uptane/commit/79a5dae1a4f91ec6ba59143163064838485e68ed)
- [Commit handling Uptane ECU Manifests](https://github.com/uptane/uptane/commit/0b831083ba58f41e98e969244c07b15aeae9900e) (The one that triggered this issue in the first place.)

**Comment by awwad** [comment]
I'm leaving this open after those changes for this remaining task:
- broad audit of ASN.1 conversion code for canonicalization
